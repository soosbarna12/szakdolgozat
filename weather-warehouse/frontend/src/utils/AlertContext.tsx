import React, { createContext, useState, useContext, ReactNode } from "react";
import { Snackbar, Alert } from "@mui/material";

type AlertContextType = {
    showAlert: (message: string, severity?: "error" | "info" | "success" | "warning") => void;
};

export const AlertContext = createContext<AlertContextType>({
    showAlert: () => { },
});

export function AlertProvider({ children }: { children: ReactNode }) {
    const [open, setOpen] = useState(false);
    const [message, setMessage] = useState("");
    const [severity, setSeverity] = useState<"error" | "info" | "success" | "warning">("info");

    function showAlert(newMessage: string, newSeverity: "error" | "info" | "success" | "warning" = "info") {
        setMessage(newMessage);
        setSeverity(newSeverity);
        setOpen(true);
    }

    return (
        <AlertContext.Provider value={{ showAlert }}>
            {children}
            <Snackbar open={open} autoHideDuration={4000} onClose={() => setOpen(false)}>
                <Alert severity={severity} onClose={() => setOpen(false)}>
                    {message}
                </Alert>
            </Snackbar>
        </AlertContext.Provider>
    );
}

export const useAlert = () => useContext(AlertContext);