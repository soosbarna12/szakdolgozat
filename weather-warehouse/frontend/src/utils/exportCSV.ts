import { download, generateCsv, mkConfig } from "export-to-csv";
import { HistoricalDataTable } from "../types/historicalDataTable.type";

export function exportCSV(data: HistoricalDataTable[]) {
  if (!data || data.length === 0) {
    return;
  }

  const csvConfig = mkConfig({ useKeysAsHeaders: true, filename: "historical_data" });

  const csv = generateCsv(csvConfig)(data as any[]);
  download(csvConfig)(csv);
}