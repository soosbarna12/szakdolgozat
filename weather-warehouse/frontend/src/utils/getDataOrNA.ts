export function getDataOrNA(data: any) {
  if (data) {
    return data;
  }
  return "N/A";
}