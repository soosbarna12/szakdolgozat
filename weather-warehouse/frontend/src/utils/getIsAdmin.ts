import { jwtDecode } from "jwt-decode";
import { JwtPayload } from "../types/jwtPayload";


export function getIsAdmin(): boolean {
    const token = localStorage.getItem('token');
    if (!token) return false;
    try {
      const decoded = jwtDecode<JwtPayload>(token);
      return decoded.role === 'admin';
    } catch {
      return false;
    }
  }