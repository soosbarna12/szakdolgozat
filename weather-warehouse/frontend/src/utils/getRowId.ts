import { HistoricalDataTable } from "../types/historicalDataTable.type";

export function getRowId(row: HistoricalDataTable): string {
  return row.date + row.cityName + row.countryCode;
}