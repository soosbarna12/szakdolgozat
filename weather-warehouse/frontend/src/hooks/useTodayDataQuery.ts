import { useQuery } from "@tanstack/react-query";
import { useAlert } from "../utils/AlertContext";
import axios from "axios";


export function useTodayDataQuery(lat: number, lon: number) {
  const { showAlert } = useAlert();
  
  const { data, error, isLoading } = useQuery({
    queryKey: ["todayWeatherData", lat, lon],
    queryFn: async () => {
      const response = await axios.get(`/today/locationData`, {
        //params: { location, lang: "en" },
        params: { lat, lon, lang: "en" },
      });

      if (response.data?.cod === "404") {
        throw new Error("City not found");
      }

      return response.data;
    },
    enabled: !!(lat && lon)
  });

  if (error) {
    showAlert("Error fetching today's weather data", "error");
  }

  return { data, error, isLoading };
}
