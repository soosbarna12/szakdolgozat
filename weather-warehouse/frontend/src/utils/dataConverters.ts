import dayjs from "dayjs";
import { HistoricalDataTable, ServerHistoricalData } from "../types/historicalDataTable.type";
import { HistoricalLocationData } from "../contexts/HistoricalContext/HistoricalContext.type";
import { getDataOrNA } from "./getDataOrNA";

export function convertKelvinToCelsius(kelvin: number | null): number | string {
	if (kelvin === null) {
		return 'N/A';
	}
	return Math.round(kelvin - 273.15);
}

export function convertKelvinToFahrenheit(kelvin: number | null): number | string {
	if (kelvin === null) {
		return 'N/A';
	}
	return Math.round(((kelvin - 273.15) * 9) / 5 + 32);
}

export function convertCelsiusToFahrenheit(celsius: number | null): number | string {
	if (celsius === null) {
		return 'N/A';
	}
	return Math.round((celsius * 9) / 5 + 32);
}

export function convertFahrenheitToCelsius(fahrenheit: number | null): number | string {
	if (fahrenheit === null) {
			return 'N/A';
	}
	return Math.round((fahrenheit - 32) * 5 / 9);
}

export function convertTitleCase(str: string): string {
	return str.charAt(0).toUpperCase() + str.slice(1).toLowerCase();
}

export function convertServerHistoricalData(historicalData: ServerHistoricalData[], location: HistoricalLocationData): HistoricalDataTable[] {
	return historicalData.map((data) => ({
		lat: location.lat,
		lon: location.lon,
		locationKey: data.LocationKey,
		dateKey: data.DateKey,
		date: dayjs(data.FullDate).format("YYYY-MM-DD"),
		cityName: data.CityName,
		countryCode: data.CountryCode,
		temp: getDataOrNA(data.Temperature),
		maxTemp: getDataOrNA(data.MaxTemperature),
		minTemp: getDataOrNA(data.MinTemperature),
		humidity: getDataOrNA(data.Humidity),
		windSpeed: getDataOrNA(data.WindSpeed),
		precipitation: getDataOrNA(data.Precipitation),
		pressure: getDataOrNA(data.Pressure),
		cloudCover: getDataOrNA(data.CloudCover),
	}));
}
