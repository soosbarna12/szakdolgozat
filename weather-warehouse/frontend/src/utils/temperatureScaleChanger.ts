import { LOCAL_STORAGE_TEMPERATURE_SCALE } from "../consts/temperatureScale.const";
import { TemperatureScale } from "../types/temperatureScale.type";
import { convertCelsiusToFahrenheit, convertFahrenheitToCelsius, convertKelvinToCelsius, convertKelvinToFahrenheit } from "./dataConverters";


export function temperatureScaleChanger(currentTempScale: TemperatureScale, goalTempScale: TemperatureScale, tempValue: number | string) : string | number {

  if (typeof tempValue === "string") {
    return tempValue;
  }

  // c to c / f to f
  if (goalTempScale === currentTempScale) {
    return tempValue;
  }
  if (goalTempScale === TemperatureScale.Celsius && currentTempScale === TemperatureScale.Fahrenheit) {
    return convertFahrenheitToCelsius(tempValue);
  }
  if (goalTempScale === TemperatureScale.Fahrenheit && currentTempScale === TemperatureScale.Celsius) {
    return convertCelsiusToFahrenheit(tempValue);
  }
  if (goalTempScale === TemperatureScale.Celsius && currentTempScale === TemperatureScale.Kelvin) {
    return convertKelvinToCelsius(tempValue);
  }
  if (goalTempScale === TemperatureScale.Fahrenheit && currentTempScale === TemperatureScale.Kelvin) {
    return convertKelvinToFahrenheit(tempValue);
  }

  return "N/A";
}
