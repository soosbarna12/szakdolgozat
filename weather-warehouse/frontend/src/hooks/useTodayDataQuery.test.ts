import '@testing-library/jest-dom';
import { waitFor } from '@testing-library/react';
import { useAlert } from '../utils/AlertContext';
import axios from '../utils/axiosConfig';
import { renderHookWithQueryClient } from '../utils/test/renderHookWithQueryClient';
import { useTodayDataQuery } from './useTodayDataQuery';

jest.mock('../utils/AlertContext', () => ({
  useAlert: jest.fn(),
}));

jest.mock('../utils/axiosConfig');

describe('useTodayDataQuery', () => {
  const mockShowAlert = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
    (useAlert as jest.Mock).mockImplementation(() => ({
      showAlert: mockShowAlert,
    }));
  });

  it('matches useTodayDataQuery hook snapshot', async () => {
    (axios.get as jest.Mock).mockReturnValue(Promise.resolve({ data: { temperature: 25 } }));

    const { result } = renderHookWithQueryClient(() => useTodayDataQuery('New York'));

    await waitFor(() => {
      expect(result.current.data).toEqual({ temperature: 25 });
    });

    expect(result.current).toMatchSnapshot();
  });

  it('fetches today’s weather data successfully', async () => {
    const mockWeatherData = { temperature: 25, humidity: 60 };
    (axios.get as jest.Mock).mockReturnValue(Promise.resolve({ data: mockWeatherData }));

    const { result } = renderHookWithQueryClient(() => useTodayDataQuery('New York'));

    await waitFor(() => {
      expect(result.current.data).toEqual(mockWeatherData);
      expect(result.current.isLoading).toBe(false);
      expect(result.current.error).toBeNull();
    });
  });

  it('shows error alert when fetching today’s weather data fails', async () => {
    (axios.get as jest.Mock).mockReturnValue(Promise.reject(new Error("Error fetching today's weather data")));

    const { result } = renderHookWithQueryClient(() => useTodayDataQuery('New York'));

    await waitFor(() => {
      expect(mockShowAlert).toHaveBeenCalledWith("Error fetching today's weather data", 'error');
      expect(result.current.error).not.toBeNull();
    });
  });

  it('throws an error when city is not found', async () => {
    (axios.get as jest.Mock).mockReturnValue(Promise.resolve({ data: { code: '404' } }));

    const { result } = renderHookWithQueryClient(() => useTodayDataQuery('InvalidCity'));

    await waitFor(() => {
      expect(result.current.error).not.toBeNull();
    });

    expect(mockShowAlert).toHaveBeenCalledWith("Error fetching today's weather data", 'error');
  });
});