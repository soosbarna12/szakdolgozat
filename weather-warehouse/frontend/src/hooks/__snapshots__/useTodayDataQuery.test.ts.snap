// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useTodayDataQuery matches useTodayDataQuery hook snapshot 1`] = `
{
  "data": {
    "temperature": 25,
  },
  "error": null,
  "isLoading": false,
}
`;
